perf: do not create a new parser for each column

olavloite opened the pull request:
Convert a Spanner value directly to the PostgreSQL equivalent without creating a Parser object. This reduces the amount of memory that is allocated and must be garbage collected for each row that is sent to the PostgreSQL client.

olavloite committed:
perf: do not create a new parser for each column

Convert a Spanner value directly to the PostgreSQL equivalent without
creating a Parser object. This reduces the amount of memory that is
allocated and must be garbage collected for each row that is sent to
the PostgreSQL client.

gauravsnj commented on src/main/java/com/google/cloud/spanner/pgadapter/parsers/BooleanParser.java:
```diff
@@ -34,6 +36,15 @@
+
+  static {
+    ByteConverter.bool(TRUE_VALUE_BYTES_BINARY, 0, true);
+    ByteConverter.bool(FALSE_VALUE_BYTES_BINARY, 0, true);
```
Why are we passing `true` as the third parameter here? Shouldn't it be `false`?

olavloite replied:
Great catch, thanks! That's absolutely right, and it was not caught by our tests for two reasons:
1. Most drivers will never use the binary format for booleans, because it does not really save any network traffic anyways (boolean values in text format are represented by `t` and `f`, so they take up just one byte, just as the binary format of 1 and 0)
2. Our tests mostly used `true` as the test value, which meant that they did not catch the fact that `false` was encoded as true.

I've added additional tests that use randomly generated values so we can catch this kind of errors in the future. That testing also turned out to detect a bug in the PostgreSQL JDBC driver, as that driver does not correctly decode binary boolean values. I've filed an issue and a PR for the JDBC driver in their repository for it.

gauravsnj commented on src/main/java/com/google/cloud/spanner/pgadapter/parsers/DateParser.java:
```diff
@@ -114,13 +116,28 @@ protected byte[] binaryParse() {
 
+  public static byte[] convertToPG(ResultSet resultSet, int position, DataFormat format) {
+    switch (format) {
+      case SPANNER:
```
Why haven't we added the conversion logic for Cloud Spanner format for some of the data types?

olavloite replied:
In Java, case statements automatically fall through, so here the same logic is used for both when the format is `SPANNER` and `POSTGRESQL_TEXT`. This is different from some other programming languages where this would mean that there is no conversion for `SPANNER`.

gauravsnj commented on src/main/java/com/google/cloud/spanner/pgadapter/parsers/DateParser.java:
```diff
@@ -114,13 +116,28 @@ protected byte[] binaryParse() {
 
+  public static byte[] convertToPG(ResultSet resultSet, int position, DataFormat format) {
+    switch (format) {
+      case SPANNER:
```
If the conversion to Spanner format is not possible for these data types, can we add some more specific error message rather than falling back to `unknown data format`?

olavloite replied:
(See above)
The default case should not be reachable, but is required for the Java compiler, as it otherwise thinks that there are code paths possible that will not return a value for this method.

gauravsnj reviewed (approved):
LGTM with some small queries

olavloite committed:
fix: numeric + bool parsing

Fix numeric parsing for NaN values.
Fix bool binary parsing (always returned true).
Add tests to verify that we actually receive the values we expect, both
when using text format and when using binary format.

olavloite committed:
Merge branch 'postgresql-dialect' into no-parser-construction

olavloite committed:
test: add missing test case
